src/Parsers/IAST.cpp

```diff
@@ -28,34 +28,86 @@ const char * IAST::hilite_none         = "\033[0m";
 
 IAST::~IAST()
 {
-    /// If deleter was set, move our children to it.
-    /// Will avoid recursive destruction.
-    if (deleter)
+    /** Create intrusive linked list of children to delete.
+      * Each ASTPtr child contains pointer to next child to delete.
+      */
+    ASTPtr delete_list_head_holder = nullptr;
+    bool delete_directly = next_to_delete_list_head == nullptr;
```

Comment:
Looks like this can be `const bool` just for safety

Reply:
Agreed. Updated.